PSA: Specification for random generation and entropy drivers

gilles-peskine-arm opened the pull request:
Create an interface for PSA random generation and entropy drivers.

stevew817 commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -345,6 +347,90 @@ Transparent drivers may provide the following key management entry points:
+
+* The random output must be of cryptographic quality, with a uniform distribution. Therefore, if the random generator includes an entropy source, this entropy source must be fed through a CSPRNG (cryptographically secure pseudo-random number generator).
+* Random generation is expected to be fast. (If a device can provide entropy but is slow at generating random data, declare it as an [entropy driver](#entropy-collection-entry-point) instead.)
+* The random generator must be able to incorporate entropy provided by an outside source.
```
What of the cases where a hardware RNG doesn't accept external input?

gilles-peskine-arm replied:
In general, that driver would only be suitable as one entropy source. It's a matter of security design: if there are multiple entropy sources, the RNG should remain strong even if all but one sources are compromised.

I do think there is a fairly common design space where there is a single peripheral that includes an entropy source and a DRBG, and the integrator does not wish to save entropy to storage. In this case, `add_entropy` is not needed.

Looking at what I've written, I think that `add_entropy` can be made optional, but a driver without an `add_entropy` entry point will be less portable: it can only be integrated on platforms without other entropy sources and without saved entropy.

stevew817 replied:
Can you clarify that in the spec, then? It currently doesn't cater to those devices that have a single entropy source and CSPRNG prepackaged in a hardware peripheral, with no way of accessing these separately.

gilles-peskine-arm replied:
Done.

stevew817 commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -345,6 +347,90 @@ Transparent drivers may provide the following key management entry points:
+* `"init_random"` (optional): if this function is present, [the core calls it once](#random-generator-initialization) after allocating a `"random_context_t"` object.
+* `"add_entropy"` (entry point): the core calls this function to [inject entropy](#entropy-injection).
+* `"get_random"` (entry point): the core calls this function whenever it needs to [obtain random data](#the-get_random-entry-point).
+* `"initial_entropy_size"` (integer): the minimum number of bytes of entropy that the core must supply before the driver can output random data.
```
Is this allowed to be 0?

gilles-peskine-arm replied:
Yes. I'll clarify.

stevew817 commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -345,6 +347,90 @@ Transparent drivers may provide the following key management entry points:
+* `"add_entropy"` (entry point): the core calls this function to [inject entropy](#entropy-injection).
+* `"get_random"` (entry point): the core calls this function whenever it needs to [obtain random data](#the-get_random-entry-point).
+* `"initial_entropy_size"` (integer): the minimum number of bytes of entropy that the core must supply before the driver can output random data.
+* `"reseed_entropy_size"` (integer): the minimum number of bytes of entropy that the core must supply when the driver runs out of entropy.
```
Is this allowed to be 0?

gilles-peskine-arm replied:
I don't think it makes sense for this to be 0. Either the driver never runs out of entropy and this value is not meaningful (although it can be a hint if the core wants to enforce prediction resistance), or the driver can run out of entropy (because its security policy treats entropy as consumable, which is mandated by some security standards) and then 0 is not valid.

stevew817 commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -191,6 +191,8 @@ The signature of a driver entry point generally looks like the signature of the
 
 Some entry points are grouped in families that must be implemented as a whole. If a driver supports an entry point family, it must provide all the entry points in the family.
 
+Drivers can also have entry points related to random generation. For transparent drivers, these are [random generation entry points](#random-generation-entry-points). For opaque drivers, these are [entropy collection entry points](#entropy-collection-entry-point).
```
Why the split? Neither of these have anything to do with key material, so shouldn't both be in the realm of transparent drivers?

gilles-peskine-arm replied:
Good point. It's closer to opaque drivers because you can have multiple entropy sources, whereas there is a single random generator. But it's only a very minor complication to allow entropy sources in transparent drivers.

stevew817 commented on include/psa/crypto_platform.h:
```diff
@@ -80,4 +80,10 @@ static inline int mbedtls_key_owner_id_equal( mbedtls_key_owner_id_t id1,
 
+#if defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+typedef struct {
+    uint8_t opaque[32];
```
isn't this very inflexible for something that is dependant on the underlying driver?

gilles-peskine-arm replied:
Yes. This is a temporary place until we add support `crypto_drivers.h`. Once we finish implementing driver support, `mbedtls_psa_external_random_context_t` and `mbedtls_psa_external_get_random` will be generated from the driver description (although we may keep the ability to define the function manually until Mbed TLS 3.0 for backward compatibility). In the meantime, I don't know what to put here as a sample type. To keep the API simple, Mbed TLS allocates the context, so it needs to know the size.

gilles-peskine-arm replied:
It's not ideal to have to define this type, but I don't see a better solution that can be done quickly.

stevew817 commented on include/mbedtls/check_config.h:
```diff
@@ -604,6 +605,11 @@
 #endif
 
+#if defined(MBEDTLS_PSA_INJECT_ENTROPY) &&              \
+    !defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
```
I think you have inverted the logic here?

gilles-peskine-arm replied:
Addressed in https://github.com/ARMmbed/mbedtls/pull/3895.

stevew817 commented on include/mbedtls/config.h:
```diff
@@ -1337,6 +1337,36 @@
+ * of random data and set \c *output_length to \p output_size.
+ *
+ * Requires: MBEDTLS_PSA_CRYPTO_C
+ *
```
Would it be good to add a warning saying that this needs to be a cryptographic-quality RNG, since it will be used for generating keys?

gilles-peskine-arm replied:
Indeed. Addressed in https://github.com/ARMmbed/mbedtls/pull/3895.

stevew817 commented:
Maybe an additional changelog entry for `MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG`?

gilles-peskine-arm commented:
To facilitate reviewing, I've split this pull request into specification and implementation. The previous version e618877ecfdb1f98f47edfab4d54c8253c058e74 is at https://github.com/gilles-peskine-arm/mbedtls/tree/psa-random-driver-spec-3.

* The new version of this PR b267d208cabe5c72084726c05afaa90c6fa5cf98 has the commits that change `docs/proposed/psa-driver-interface.md`.
* https://github.com/ARMmbed/mbedtls/pull/3895 has the code. Initially, at commit 25c8deb06c0a4e1fcdf09fb71366015812a8e96e, this is the same code that was posted here in e618877ecfdb1f98f47edfab4d54c8253c058e74.

mpg reviewed (commented):
This is looking pretty good! Other than two language issues, my only real concern is about `initial_entropy_size` defaulting to `0`.

gilles-peskine-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -345,6 +390,99 @@ Transparent drivers may provide the following key management entry points:
+* `"init_random"` (entry point, optional): if this function is present, [the core calls it once](#random-generator-initialization) after allocating a `"random_context_t"` object.
+* `"add_entropy"` (entry point, optional): the core calls this function to [inject entropy](#entropy-injection). This entry point is optional if the driver is for a peripheral that includes an entropy source of its own, however [random generator drivers without entropy injection](#random-generator-drivers-without-entropy-injection) have limited portability since they can only be used on platforms ...
+* `"get_random"` (entry point, mandatory): the core calls this function whenever it needs to [obtain random data](#the-get_random-entry-point).
+* `"initial_entropy_size"` (integer, optional): the minimum number of bytes of entropy that the core must supply before the driver can output random data. This can be `0` if the driver is for a peripheral that includes an entropy source of its own. If omitted, the value is `0`.
```
This value comes from the driver. If the device includes an entropy source, it doesn't need external entropy, so it should declare `initial_entropy_size` = 0. It still makes sense to have a nonzero `reseed_entropy_size` to convey “if you're going to reseed anyway, this is how many bytes I expect”.

If there are other entropy sources on the device, the core would call `add_entropy` initially, even if `initial_entropy_size` is 0.

mpg replied:
> If the device includes an entropy source, it doesn't need external entropy, so it should declare `initial_entropy_size` = 0

Yes, I have no problem with the driver being able to declare that explicitly, I'm more concerned about making it the default. 

> If there are other entropy sources on the device, the core would call `add_entropy` initially, even if `initial_entropy_size` is 0.

Ok, so apparently I don't understand what `initial_entropy_size == 0` means. I thought it meant "dear core, please don't bother feeding me entropy before calling `get_random`". Now you're telling me the core will call `add_entropy` anyway - how many times will it call it? up to a total of how many bytes of entropy?

gilles-peskine-arm replied:
It's not exactly “don't bother feeding me entropy”, but rather “you don't need to feed me entropy”.

Maybe it would be better to have a separate boolean parameter `needs_initial_entropy`? But I think it would be redundant with `initial_entropy_size` being 0.

mpg replied:
Ok, thanks for clarifying the meaning of `initial_entropy_size == 0`. I don't think an extra parameter is required. My main question is really about the default value, and I don't think you've answered that so far.

Defaulting to `0` means "a driver doesn't need entropy unless it explicitly declared it needed it". It would seem much safer to me to take the approach that "a driver needs entropy unless it explicitly declared it didn't".

By defaulting to "doesn't need entropy" we run the risk of silently getting non-random bytes if the driver forgot to declare an `initial_entropy_size`. If we defaulted to "the expected security strength of the device" I don't see what serious issue it would cause: if the driver forgets to declare it doesn't need entropy and there are no other entropy sources, then we would get an error, which would be easily fixed by declaring an `initial_entropy_size` of `0`.

gilles-peskine-arm replied:
Even if `initial_entropy_size` is 0, the core is allowed to make an initial call `add_entropy`. The “security strength of the device” is a policy choice, not something that this specification mandates. So defaulting to “the expected security strength of the device” would mean defaulting to a value chosen by the core anyway.

mpg replied:
I understand that the core is allowed to make an initial call to `add_entropy`, with an amount of entropy of its choosing. The question is what happens when it can't, because it has no entropy sources. I think we can agree that in this case, either the driver has its own source of entropy, or we'd like to give an error (rather than output predictable "random" bytes).

My point is that we shouldn't just assume that a driver has its own source of entropy: we should require the driver to declare it explicitly, and I think *omitting* a parameter is much less explicit than setting it to `0`.

Or maybe we can turn the question: why do you think `0` is better than other choices for the default value?

gilles-peskine-arm replied:
Fair enough, then I guess the parameter should be mandatory (unless `add_entropy` is omitted).

mpg replied:
Can we make it always mandatory and it must be `0` is `add_entropy` is omitted? That would provide an explicit confirmation that `add_entropy` was omitted on purpose rather than by accident.

gilles-peskine-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -345,6 +390,99 @@ Transparent drivers may provide the following key management entry points:
+* `context`: a random generation context. On the first call to `"add_entropy"`, this object has been initialized by a call to the driver's `"init_random"` entry point if one is present, and to all-bits-zero otherwise.
+* `entropy`: a buffer containing high-entropy data to seed the random generator.
+* `entropy_size`: the size of the `entropy` buffer in bytes. It is guaranteed to be at least `1`, but it may be smaller than the amount of entropy that the driver needs to deliver random data, in which case the core will call the `"add_entropy"` entry point again to supply more entropy.
+
```
This is an RNG driver, not just a PRNG driver. Most peripherals that have PRNG acceleration also include a TRNG. Accelerators without a TRNG usually only accelerate something more basic, e.g. only AES and not AES+CTR_DRBG.

The core manages total entropy estimates, but only the driver can estimate the amount of entropy returned by the TRNG.

gilles-peskine-arm replied:
There's an open question whether `entropy` should be full-entropy, or whether the function should take a separate parameter to convey the entropy density. I agree that “high-entropy” is unclear though.

mpg replied:
Yes, I saw the open question later an forgot to update this comment.

mpg commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -756,6 +756,19 @@ psa_set_key_lifetime(&attributes, PSA_KEY_LIFETIME_FROM_PERSISTENCE_AND_LOCATION
+
+It would be better if there was a uniform requirement on integer values. Do they have to be JSON integers? C preprocessor integers (which could be e.g. a macro defined in some header file)? C compile-time constants (allowing `sizeof`)?
+
+This choice is partly driver by the use of the values, so they might not be uniform. Note that if the value can be zero and it's plausible that the core would want to statically allocate an array of the given size, the core needs to know whether the value is 0 so that it could use code like
```
Typo: s/driver/driven/

gilles-peskine-arm replied:
Fixed by amending the last commit.

mpg reviewed (approved):
Thanks for addressing my concerns about `initial_entropy_size`'s default value. Looks good to me now.

mpg reviewed (dismissed):
LGTM

ronald-cron-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -683,6 +834,26 @@ An example use case for updating the persistent state at arbitrary times is to r
+
+#### Random generator instantiations
+
+May the core instantiate a random generation context more than once? In other words, can there be multiple objects of type `acme_random_context_t`?
```
Yes I was wondering about that when reading as well.

ronald-cron-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -683,6 +834,26 @@ An example use case for updating the persistent state at arbitrary times is to r
+
+May the core instantiate a random generation context more than once? In other words, can there be multiple objects of type `acme_random_context_t`?
+
+Functionally, one RNG is as good as any. If the core wants some parts of the system to use a deterministic generator for reproducibility, it can't use this interface anyway, since the RNG is not necessarily deterministic. However, for performance on multiprocessor systems, a multithreaded core could prefer to use one RNG instance per thread.
```
> However, for performance on multiprocessor systems, a multithreaded core could prefer to use one RNG instance per thread

But at the end there is only one hardware anyway that threads will compete for thus not sure why it would be more performant.

gilles-peskine-arm replied:
Not necessarily: in a multiprocessor system, there may well be an RNG per core.

ronald-cron-arm replied:
Ah right, that's indeed the case I think with recent x86 processors.

ronald-cron-arm reviewed (commented):
I have a few questions and suggestions.

gilles-peskine-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -345,6 +390,99 @@ Transparent drivers may provide the following key management entry points:
+
+* `context`: a random generation context. On the first call to `"add_entropy"`, this object has been initialized by a call to the driver's `"init_random"` entry point if one is present, and to all-bits-zero otherwise.
+* `entropy`: a buffer containing full-entropy data to seed the random generator. “Full-entropy” means that the data is uniformly distributed and independent of any other observable quantity.
+* `entropy_size`: the size of the `entropy` buffer in bytes. It is guaranteed to be at least `1`, but it may be smaller than the amount of entropy that the driver needs to deliver random data, in which case the core will call the `"add_entropy"` entry point again to supply more entropy.
```
Indeed, an output of `get_entropy` is generally not suitable as an input for `add_entropy`. But I don't think this is the right place to elaborate on that: mixing the get-random interface and the get-entropy interface could get confusing quickly. I'll add a separate subsection on this topic.

gilles-peskine-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -345,6 +390,99 @@ Transparent drivers may provide the following key management entry points:
+* After a call to the `"get_random"` entry point returns less than the required amount of random data, to supply `"reseed_entropy_size"` bytes of entropy.
+
+When the driver requires entropy, the core can supply it with one or more successive calls to the `"add_entropy"` entry point. If the required entropy size is zero, the core does not need to call `"add_entropy"`.
+
```
I don't have anything to say about status codes for `init_random` or `add_entropy`. As usual, `PSA_SUCCESS` for success, or an applicable error code.

gilles-peskine-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -683,6 +834,26 @@ An example use case for updating the persistent state at arbitrary times is to r
+
+* Raw data: drivers must implement entropy mixing. `"add_entropy"` needs an extra parameter to indicate the amount of entropy in the data. The core must not do any conditioning.
+* Choice: drivers must implement entropy mixing. `"add_entropy"` needs an extra parameter to indicate the amount of entropy in the data. The core may do conditioning if it wants, but doesn't have to.
+* Full entropy: drivers don't need to do entropy mixing.
```
Not very usefully. It typically uses building blocks (such as a hash or MAC or XOF algorithm) that can be accelerated, but it doesn't have a standard interface and there'd be little advantage in terms of code size or speed.

ronald-cron-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -313,6 +315,47 @@ TODO
+
+#### Entropy collection and blocking
+
+The intent of the `BLOCK` and `KEEPALIVE` [flags](#entropy-collection-flags) is to support drivers for TRNG (True Random Number Generator, i.e. an entropy source peripheral) that have a long ramp-up time, especially on platforms with multiple entropy sources.
```
Yes the suspend/shutdown of the TRNG would be driven by the software driver but if the hardware TRNG cannot be suspend/shutdown (always on device) while the system remain up there is nothing the software driver can do. Thus, to me, the KEEPALIVE flag is useful if the TRNG has a long ramp-up time and can be suspended/shut down.

ronald-cron-arm replied:
I would add a sentence like "The `KEEPALIVE` flag is to support drivers for TRNG that additionally can be suspended or shut down to save power." to clarify the intent of the `KEEPALIVE` flag.

gilles-peskine-arm replied:
I added a somewhat different sentence with the same intent.

ronald-cron-arm commented on docs/proposed/psa-driver-interface.md:
```diff
@@ -449,6 +449,16 @@ The core may call this function at any time. For example, to enforce prediction
+
+Note that `"get_entropy"` returns data with an estimated amount of entropy that is in general less than the buffer size. The core must apply a mixing algorithm to the output of `"get_entropy"` to obtain full-entropy data.
+
+For example, the core may use a simple mixing scheme based on a pseudorandom function family $(F_k)$ with an $E$-bit output where $E = 8 \cdot \mathtt{entropy_size}$ and $\mathtt{entropy_size}$ is the desired amount of entropy in bytes (typically the random driver's `"initial_entropy_size"` property for the initial seeding and the `"reseed_entropy_size"` property for subsequent reseeding). The co ...
```
Just checking but with the rich diff "$E = 8 \cdot \mathtt{entropy_size}$ and $\mathtt{entropy_size}$" is also rendered as it is and I assume this is not the intent. Is this eventually rendered "properly" on GitHub?

gilles-peskine-arm replied:
No, GitHub doesn't render math. You have to use a different markdown renderer such as pandoc.

ronald-cron-arm reviewed (commented):
Thanks for the changes and improvements. I am almost done with the review, just three last things: two pending comments from the previous review and on new comment.

gilles-peskine-arm commented:
As in https://github.com/ARMmbed/mbedtls/pull/3878#issuecomment-739991534, there was a merge conflict due to the time stamp from a parallel edit, so I've moved the base to current `development`, removed the time stamp (using the same commit as on #3878) and rebased all the existing commits on top of the new base. The previous version is at https://github.com/gilles-peskine-arm/mbedtls/tree/psa-random-driver-spec-4. I'll address the remaining review comments in subsequent commits.

gilles-peskine-arm commented:
@ronald-cron-arm  Do the remaining threads (https://github.com/ARMmbed/mbedtls/pull/3882#discussion_r523970675, https://github.com/ARMmbed/mbedtls/pull/3882#discussion_r531458257, https://github.com/ARMmbed/mbedtls/pull/3882#discussion_r535007295) require action from me? I read them as questions and I don't think they justify changing the document, but as I understand it you aren't happy with the current state of the document, so can you please clarify what kind of changes you want in each case?

ronald-cron-arm commented:
I've proposed a [change](https://github.com/ARMmbed/mbedtls/pull/3882/files#r538260432) for one of the three comments that were still not resolved and resolved the two other ones.

gilles-peskine-arm commented:
Given that Manuel and Ronald are at most requesting minor changes, and there is no opposing voice, I consider this draft approved for design. I will circulate it more widely, especially among TRNG makers, before considering the design to be fully stable.

mpg reviewed (approved):
I reviewed the changes since my previous approval and found nothing wrong.

ronald-cron-arm reviewed (approved):
LGTM as well.
